### PR TITLE
MIPS VM: Add instruction pre-fetching

### PR DESCRIPTION
**Pull request by mrmr1993**

### CODE REVIEW

File: optimism/src/mips/interpreter.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    ShiftRightLogical,            // srl
+    ShiftRightArithmetic,         // sra
+    ShiftLeftLogicalVariable,     // sllv
+    ShiftRightLogicalVariable,    // srlv
+    ShiftRightArithmeticVariable, // srav
+    JumpRegister,                 // jr
+    JumpAndLinkRegister,          // jalr
+    SyscallMmap,                  // syscall (Mmap)
+    SyscallExitGroup,             // syscall (ExitGroup)
+    SyscallReadPrimage,           // syscall (Read 5)
```
Comment (rbonichon) [inline optimism/src/mips/interpreter.rs:30]: Typo fix
```suggestion
    SyscallReadPreimage,           // syscall (Read 5)
```

File: optimism/src/mips/witness.rs [hunk 2]
```diff
@@ -92,7 +98,135 @@ impl<Fp: Field> Env<Fp> {
...
+                        Instruction::RType(RTypeInstruction::SyscallOther)
+                    }
+                    4120 => {
+                        // sysClone
+                        Instruction::RType(RTypeInstruction::SyscallOther)
+                    }
+                    4246 => Instruction::RType(RTypeInstruction::SyscallExitGroup),
+                    4003 => match self.registers.general_purpose[4] {
+                        interpreter::FD_PREIMAGE_READ => {
+                            Instruction::RType(RTypeInstruction::SyscallReadPrimage)
```
Comment (rbonichon) [inline optimism/src/mips/witness.rs:145]:
```suggestion
                            Instruction::RType(RTypeInstruction::SyscallReadPreimage)
```

**Review by rbonichon — APPROVED** [review]
LGTM!

Picked a few opcodes here and there to double-check and did not find any error. By  induction, there are no errors in the decoding. QED. 😸

**Comment by dannywillems** [comment]
Integrated the changes in https://github.com/o1-labs/proof-systems/pull/1322

**Review by dannywillems — APPROVED** [review]

**Pull request merged**